ocaml/xapi/xapi_pusb.ml

```diff
@@ -0,0 +1,91 @@
+(*
+ * Copyright (C) 2006-2009 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Stdext
+open Listext
+open Threadext
+module D = Debug.Make(struct let name="xapi" end)
+open D
+
+let mutex = Mutex.create ()
+let create ~__context ~uSB_group ~host ~other_config ~path
+      ~vendor_id ~vendor_desc ~product_id ~product_desc ~serial ~version ~description =
+  let pusb = Ref.make () and uuid = Uuid.make_uuid () in
+  let host = Helpers.get_localhost ~__context in
+  Mutex.execute mutex (fun () ->
+      Db.PUSB.create ~__context ~ref:pusb ~uuid:(Uuid.to_string uuid)
+        ~uSB_group ~host ~attached:Ref.null ~other_config ~path ~vendor_id ~vendor_desc ~product_id
+         ~product_desc ~serial ~version ~description ~passthrough_enabled:false;
+  );
+  debug "PUSB ref='%s' created" (Ref.string_of pusb);
+  pusb
+
+let scan_start ~__context =
+  let host = Helpers.get_localhost ~__context in
+  let usbs = Xapi_pusb_helpers.get_usbs ~__context in
+  let existing_pusbs = List.filter (fun (rf, rc) -> rc.API.pUSB_host = host) (Db.PUSB.get_all_records ~__context) in
+  let rec find_or_create cur = function
+    | [] -> cur
+    | usb :: remaining_usbs ->
+      let pusb =
+        try
+          let (rf, rc) = List.find (fun (_, rc) -> Xapi_pusb_helpers.is_same_pusb rc usb) existing_pusbs
+          in
+          (rf, rc)
+        with Not_found ->
+          let path = List.hd (Xapi_pusb_helpers.extract_member usb "path") in
```

Comment:
Can you be sure these `List.hd` operations never fails because the list is never empty?

Reply:
If usbs is empty, will not get here.  when usbs is not empty, the element of usb( eg. [{"product-desc": "test", "product-id": "3002", "description": "14d6_3002", "vendor-desc": "test", "version": "1.10", "vendor-id": "14d6", "path": "1-4", "serial": ""}] will always exists. So I think there's no need to  throw failure here. How do you think so?